File: sdk/python/aragora_sdk/namespaces/slo.py
"""
SLO (Service Level Objective) Namespace API

Provides methods for monitoring SLO compliance, error budgets, violations, and alerts:
- Get overall SLO status
- Get individual SLO details (availability, latency, debate-success)
- Monitor error budget consumption
- List and track violations
- Check compliance status
- View active alerts
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class SLOAPI:
    """
    Synchronous SLO (Service Level Objective) API.

    Provides methods for monitoring service level objectives including
    compliance status, error budgets, violations, and alerts.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> status = client.slo.get_status()
        >>> print(f"Overall status: {status['status']}")
        >>> budget = client.slo.get_error_budget()
        >>> print(f"Error budget remaining: {budget['remaining_percent']}%")
        >>> violations = client.slo.get_violations()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # =========================================================================
    # Overall Status
    # =========================================================================

    def get_status(self) -> dict[str, Any]:
        """
        Get overall SLO compliance status.

        Returns status of all SLOs including compliance percentages,
        summary statistics, and active alerts.

        Returns:
            Dict with overall SLO status including:
            - status: Overall health ('healthy', 'degraded', 'critical')
            - timestamp: Status timestamp
            - slos: Individual SLO statuses
            - alerts: Active alerts
            - summary: Compliance summary statistics
        """
        return self._client.request("GET", "/api/v2/slo/status")

    # =========================================================================
    # Individual SLOs
    # =========================================================================

    def get_availability(self) -> dict[str, Any]:
        """
        Get availability SLO status.

        Returns:
            Dict with availability SLO metrics including current value,
            target, and compliance status.
        """
        return self._client.request("GET", "/api/slos/availability")

    def get_latency(self) -> dict[str, Any]:
        """
        Get latency SLO status.

        Returns:
            Dict with latency SLO metrics including P50, P95, P99 values.
        """
        return self._client.request("GET", "/api/slos/latency")

    def get_debate_success(self) -> dict[str, Any]:
        """
        Get debate success rate SLO status.

        Returns:
            Dict with debate success SLO metrics.
        """
        return self._client.request("GET", "/api/slos/debate-success")

    def get_slo(self, slo_name: str) -> dict[str, Any]:
        """
        Get details for a specific named SLO.

        Args:
            slo_name: SLO name (e.g., 'availability', 'latency', 'debate_success').

        Returns:
            Dict with individual SLO details including current value and target.
        """
        return self._client.request("GET", f"/api/slos/{slo_name}")

    # =========================================================================
    # Error Budget
    # =========================================================================

    def get_error_budget(self) -> dict[str, Any]:
        """
        Get error budget timeline across all SLOs.

        Returns:
            Dict with error budget data including:
            - remaining_percent: Percentage of error budget remaining
            - consumed_percent: Percentage of error budget consumed
            - timeline: Error budget consumption over time
        """
        return self._client.request("GET", "/api/slos/error-budget")

    def get_enforcer_budget(self) -> dict[str, Any]:
        """
        Get SLO enforcer error budget data.

        Returns:
            Dict with per-SLO remaining/consumed error budget from the enforcer.
        """
        return self._client.request("GET", "/api/v1/slo/budget")

    def get_debate_health(
        self,
        *,
        window: str = "1h",
        all_windows: bool = False,
    ) -> dict[str, Any]:
        """
        Get debate SLO health with optional window selection.

        Args:
            window: Time window to evaluate ("1h", "24h", "7d").
            all_windows: When true, returns multi-window health data.

        Returns:
            Dict containing debate SLO health data.
        """
        params: dict[str, Any] = {"window": window}
        if all_windows:
            params["all_windows"] = "true"
        return self._client.request("GET", "/api/health/slos", params=params)

    def get_enforcer_budget(self) -> dict[str, Any]:
        """
        Get SLO enforcer error budget data.

        Returns:
            Dict with per-SLO remaining/consumed error budget from the enforcer.
        """
        return self._client.request("GET", "/api/v1/slo/budget")

    def get_debate_health(
        self,
        *,
        window: str = "1h",
        all_windows: bool = False,
    ) -> dict[str, Any]:
        """
        Get debate SLO health with optional window selection.

        Args:
            window: Time window to evaluate ("1h", "24h", "7d").
            all_windows: When true, returns multi-window health data.

        Returns:
            Dict containing debate SLO health data.
        """
        params: dict[str, Any] = {"window": window}
        if all_windows:
            params["all_windows"] = "true"
        return self._client.request("GET", "/api/health/slos", params=params)

    # =========================================================================
    # Violations
    # =========================================================================

    def get_violations(self) -> dict[str, Any]:
        """
        Get recent SLO violations.

        Returns:
            Dict with recent violations including timestamps,
            affected SLOs, and severity.
        """
        return self._client.request("GET", "/api/slos/violations")

    # =========================================================================
    # Targets
    # =========================================================================

    def get_targets(self) -> dict[str, Any]:
        """
        Get configured SLO targets.

        Returns:
            Dict with configured target values for each SLO including
            thresholds, time windows, and alerting rules.
        """
        return self._client.request("GET", "/api/slos/targets")


class AsyncSLOAPI:
    """
    Asynchronous SLO (Service Level Objective) API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     status = await client.slo.get_status()
        ...     budget = await client.slo.get_error_budget()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # =========================================================================
    # Overall Status
    # =========================================================================

    async def get_status(self) -> dict[str, Any]:
        """Get overall SLO compliance status."""
        return await self._client.request("GET", "/api/v2/slo/status")

    # =========================================================================
    # Individual SLOs
    # =========================================================================

    async def get_availability(self) -> dict[str, Any]:
        """Get availability SLO status."""
        return await self._client.request("GET", "/api/slos/availability")

    async def get_latency(self) -> dict[str, Any]:
        """Get latency SLO status."""
        return await self._client.request("GET", "/api/slos/latency")

    async def get_debate_success(self) -> dict[str, Any]:
        """Get debate success rate SLO status."""
        return await self._client.request("GET", "/api/slos/debate-success")

    async def get_slo(self, slo_name: str) -> dict[str, Any]:
        """Get details for a specific named SLO."""
        return await self._client.request("GET", f"/api/slos/{slo_name}")

    # =========================================================================
    # Error Budget
    # =========================================================================

    async def get_error_budget(self) -> dict[str, Any]:
        """Get error budget timeline across all SLOs."""
        return await self._client.request("GET", "/api/slos/error-budget")

    async def get_enforcer_budget(self) -> dict[str, Any]:
        """Get SLO enforcer error budget data."""
        return await self._client.request("GET", "/api/v1/slo/budget")

    async def get_debate_health(
        self,
        *,
        window: str = "1h",
        all_windows: bool = False,
    ) -> dict[str, Any]:
        """Get debate SLO health with optional window selection."""
        params: dict[str, Any] = {"window": window}
        if all_windows:
            params["all_windows"] = "true"
        return await self._client.request("GET", "/api/health/slos", params=params)

    async def get_enforcer_budget(self) -> dict[str, Any]:
        """Get SLO enforcer error budget data."""
        return await self._client.request("GET", "/api/v1/slo/budget")

    async def get_debate_health(
        self,
        *,
        window: str = "1h",
        all_windows: bool = False,
    ) -> dict[str, Any]:
        """Get debate SLO health with optional window selection."""
        params: dict[str, Any] = {"window": window}
        if all_windows:
            params["all_windows"] = "true"
        return await self._client.request("GET", "/api/health/slos", params=params)

    # =========================================================================
    # Violations
    # =========================================================================

    async def get_violations(self) -> dict[str, Any]:
        """Get recent SLO violations."""
        return await self._client.request("GET", "/api/slos/violations")

    # =========================================================================
    # Targets
    # =========================================================================

    async def get_targets(self) -> dict[str, Any]:
        """Get configured SLO targets."""
        return await self._client.request("GET", "/api/slos/targets")

File: sdk/typescript/src/namespaces/slo.ts
/**
 * SLO (Service Level Objective) Namespace API
 *
 * Provides endpoints for monitoring SLO compliance, error budgets, and violations.
 */

import type { AragoraClient } from '../client';

/**
 * SLO target configuration
 */
export interface SLOTarget {
  name: string;
  target_percent: number;
  window_days: number;
  description?: string;
}

/**
 * Current SLO status
 */
export interface SLOStatus {
  name: string;
  current_percent: number;
  target_percent: number;
  is_meeting: boolean;
  window_start: string;
  window_end: string;
  total_requests: number;
  successful_requests: number;
  failed_requests: number;
}

/**
 * Error budget information
 */
export interface ErrorBudget {
  slo_name: string;
  budget_percent: number;
  consumed_percent: number;
  remaining_percent: number;
  is_exhausted: boolean;
  burn_rate: number;
  projected_exhaustion?: string;
  window_days: number;
}

/**
 * SLO violation record
 */
export interface SLOViolation {
  slo_name: string;
  timestamp: string;
  actual_percent: number;
  target_percent: number;
  duration_seconds: number;
  severity: 'warning' | 'critical';
  resolved: boolean;
  resolved_at?: string;
}

/**
 * Overall SLO compliance status
 */
export interface OverallSLOStatus {
  status: 'healthy' | 'degraded' | 'critical';
  timestamp: string;
  slos: Record<string, SLOStatus>;
  alerts: SLOAlert[];
  summary: {
    total: number;
    meeting: number;
    not_meeting: number;
    compliance_percent: number;
  };
}

/**
 * SLO alert information
 */
export interface SLOAlert {
  slo_name: string;
  severity: 'warning' | 'critical';
  message: string;
  triggered_at: string;
  acknowledged: boolean;
}

/**
 * SLO namespace for service level objective monitoring.
 *
 * @example
 * ```typescript
 * // Check overall SLO status
 * const status = await client.slo.getStatus();
 * console.log(`SLO compliance: ${status.summary.compliance_percent}%`);
 *
 * // Get error budget for availability
 * const budget = await client.slo.getErrorBudget('availability');
 * console.log(`Error budget remaining: ${budget.remaining_percent}%`);
 *
 * // List recent violations
 * const violations = await client.slo.getViolations({ limit: 10 });
 * violations.forEach(v => console.log(`${v.slo_name}: ${v.severity}`));
 * ```
 */
export class SLONamespace {
  constructor(private client: AragoraClient) {}

  /**
   * Get overall SLO compliance status.
   *
   * Returns status of all SLOs including compliance percentages and alerts.
   */
  async getStatus(): Promise<OverallSLOStatus> {
    return this.client.request<OverallSLOStatus>('GET', '/api/v1/slos/status');
  }

  /**
   * Get individual SLO details.
   *
   * @param sloName - Name of the SLO (e.g., 'availability', 'latency')
   */
  async getSLO(sloName: string): Promise<SLOStatus> {
    return this.client.request<SLOStatus>('GET', `/api/v1/slos/${sloName}`);
  }

  /**
   * Get availability SLO status.
   */
  async getAvailability(): Promise<SLOStatus> {
    return this.getSLO('availability');
  }

  /**
   * Get latency SLO status.
   */
  async getLatency(): Promise<SLOStatus> {
    return this.getSLO('latency');
  }

  /**
   * Get debate success SLO status.
   */
  async getDebateSuccess(): Promise<SLOStatus> {
    return this.getSLO('debate-success');
  }

  /**
   * Get error budget information.
   *
   * @param sloName - Optional SLO name to get specific budget
   */
  async getErrorBudget(sloName?: string): Promise<ErrorBudget | ErrorBudget[]> {
    const path = sloName
      ? `/api/v1/slos/error-budget?slo=${sloName}`
      : '/api/v1/slos/error-budget';
    return this.client.request<ErrorBudget | ErrorBudget[]>('GET', path);
  }

  /**
   * Get SLO violations.
   *
   * @param options.limit - Maximum violations to return (default: 50)
   * @param options.sloName - Filter by specific SLO
   * @param options.severity - Filter by severity
   * @param options.unresolved - Only show unresolved violations
   */
  async getViolations(options?: {
    limit?: number;
    sloName?: string;
    severity?: 'warning' | 'critical';
    unresolved?: boolean;
  }): Promise<SLOViolation[]> {
    const params = new URLSearchParams();
    if (options?.limit) params.set('limit', options.limit.toString());
    if (options?.sloName) params.set('slo', options.sloName);
    if (options?.severity) params.set('severity', options.severity);
    if (options?.unresolved) params.set('unresolved', 'true');

    const query = params.toString();
    const path = query ? `/api/v1/slos/violations?${query}` : '/api/v1/slos/violations';
    return this.client.request<SLOViolation[]>('GET', path);
  }

  /**
   * Get configured SLO targets.
   */
  async getTargets(): Promise<SLOTarget[]> {
    return this.client.request<SLOTarget[]>('GET', '/api/v1/slos/targets');
  }

  /**
   * Check if all SLOs are meeting their targets.
   */
  async isCompliant(): Promise<boolean> {
    const status = await this.getStatus();
    return status.status === 'healthy';
  }

  /**
   * Get current alerts for SLO violations.
   */
  async getAlerts(): Promise<SLOAlert[]> {
    const status = await this.getStatus();
    return status.alerts;
  }

  /**
   * Get debate SLO health for operational windows.
   *
   * @param options.window - Window to evaluate ("1h", "24h", "7d")
   * @param options.allWindows - Include all windows when true
   */
  async getDebateHealth(options?: {
    window?: '1h' | '24h' | '7d';
    allWindows?: boolean;
  }): Promise<Record<string, unknown>> {
    const params: Record<string, string> = {};
    if (options?.window) params.window = options.window;
    if (options?.allWindows !== undefined) {
      params.all_windows = String(options.allWindows);
    }

    return this.client.request<Record<string, unknown>>(
      'GET',
      '/api/health/slos',
      Object.keys(params).length > 0 ? { params } : {}
    );
  }

  /**
   * Get SLO enforcer error budget information.
   */
  async getEnforcerBudget(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/slo/budget') as Promise<Record<string, unknown>>;
  }

  /**
   * Get SLO operational status.
   */
  async getSloStatus(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/slo/status') as Promise<Record<string, unknown>>;
  }

}
